core/serialization/blocks.js

```diff
@@ -162,6 +163,11 @@ const saveExtraState = function(block, state) {
     if (extraState !== null) {
       state['extraState'] = extraState;
     }
+  } else if (block.mutationToDom) {
+    const extraState = block.mutationToDom();
```

Comment:
Can `extraState` here end up duplicating some of the non-extra state on `state`?

Reply:
It shouldn't unless developers explicitly want it to. `mutationToDom` doesn't have a default implementation. All we're getting from it is whatever developers define.

I'm worried I'm missing something here. Could you expound?

Reply:
I was just wondering if any of the blocks we provide implement both `saveExtraState` and `mutationToDom`, and so if a developer overrode (only) the with an implementation that did a super call whether this might result in the same information being recorded both in fields in the JSO (by our block's `saveExtraState`) and as XML in the JSO (via override's super call).

But it looks like if a block implements `saveExtraState` then its `mutationToDom` is completely ignored.  Is that correct?

Reply:
> But it looks like if a block implements saveExtraState then its mutationToDom is completely ignored. Is that correct?

That is correct! We ignore `mutationToDom` because we assume that if a dev has implemented the equivalent JSO hook (`saveExtraState`) they want us to use that.

> ...did a super call ...

Also, all block types are (basically) implemented as mixins, so we don't have to worry about subclassing afaik, because it's not supported. Eg the [lists_create_with](https://github.com/google/blockly/blob/f94b1db85018b5a7261fa01f7d04cd5c8367fe2c/blocks/lists.js#L118) block.

Reply:
> Also, all block types are (basically) implemented as mixins, so we don't have to worry about subclassing afaik

TIL (though possibly not for the first time…)